### PR TITLE
Fix issues of Modf

### PR DESCRIPTION
**Pull request by amdrexu**
When x is -0.0, modf will return 0.0 according to this formula:

`  modf(x) = x - trunc(x)`

This leads to an addition of (-0.0) + 0.0 and our HW returns 0.0. According to SPIR-V spec, the result has the same sign as that of the input. We must check this special case by doing this:

`  modf(x) = copysign(modf(x), x)`

Also, according to C modf function, when x=+INF/-INF, the result is 0.0 but we have such computation (-INF) + INF or INF - INF. Our HW returns NaN so we must check this special case as well by do this:

`  modf(x) = (x == INF) ? copysign(0.0, x) : modf(x)`

### CODE REVIEW

**Comment by amdrexu** [comment]
Not sure if this is a good idea to put `Modf` to LGC builder.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 0ef68649818664e72e9559753aeca90ba56f0b6e
<details><summary>CTS tests (Failed: 0/138020)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35401/69018 (51.3%)</li>
<li>Failed: 0/69018 (0.0%)</li>
<li>Not Supported: 33617/69018 (48.7%)</li><li>Warnings: 0/69018 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35424/69002 (51.3%)</li>
<li>Failed: 0/69002 (0.0%)</li>
<li>Not Supported: 33578/69002 (48.7%)</li><li>Warnings: 0/69002 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by nhaehnle** [comment]
> Not sure if this is a good idea to put `Modf` to LGC builder.

Could be useful if llpcfe needs it as well, but I don't really have an opinion on it.

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 1]
```diff
@@ -8903,25 +8903,42 @@ Value *SPIRVToLLVM::transGLSLExtInst(SPIRVExtInst *extInst, BasicBlock *bb) {
...
+    }
+
+    if (extOp == GLSLstd450Modf) {
+      // Vectors are represented as arrays in memory, so we need to cast the pointer of array to pointer of vector
+      // before storing.
+      if (wholeNum->getType()->isVectorTy()) {
+        assert(args[1]->getType()->isPointerTy());
+        Type *const castType = wholeNum->getType()->getPointerTo(args[1]->getType()->getPointerAddressSpace());
+        args[1] = getBuilder()->CreateBitCast(args[1], castType);
+      }
```
[the remark this answers is not in the document]
Comment (amdrexu) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp:8937]: Yeah. this code is legacy. I will remove it.

**Comment by amdrexu** [comment]
> > Not sure if this is a good idea to put `Modf` to LGC builder.
> 
> Could be useful if llpcfe needs it as well, but I don't really have an opinion on it.

I consult Xavi. And it seems there is no such need so I will keep it in SPIR-V frontend not LGC.

**Comment by amdvlk-admin** [comment]
### Test summary for commit c5b60394b6a72432313769c446fc3f310be11018
<details><summary>CTS tests (Failed: 0/137941)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35352/68939 (51.3%)</li>
<li>Failed: 0/68939 (0.0%)</li>
<li>Not Supported: 33587/68939 (48.7%)</li><li>Warnings: 0/68939 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35424/69002 (51.3%)</li>
<li>Failed: 0/69002 (0.0%)</li>
<li>Not Supported: 33578/69002 (48.7%)</li><li>Warnings: 0/69002 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — APPROVED** [review]

**Pull request merged**